### PR TITLE
[Feature:Submission] Gradeable Pre-Requisite Locking

### LINKED ISSUE #3698: Gradeable Locked until completion of prior gradeable
New option on  create gradeable:  "depends on" or "locked until gradeable X is completed"

Add a new field in the database for "depends on" (another valid gradeable id)

For students, the submission button will be greyed out (similar to team submissions before you have created your team) and they will be blocked from accessing page until they have full credit on the specified assignment.

### PR DESCRIPTION
**Pull request by cjreed121**
### Please check if the PR fulfills these requirements:

* [x] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Currently there is no support for automatic opening of gradeables for an individual student after they complete a different gradeable.
Closes #3698

### What is the new behavior?
Now there is a support for gradeables to be locked until a student completes the pre-requisite gradeable with a certain autograder score. Until the student completes the gradeable with a high enough score they will not be able to click or submit the next assignment. It will just say locked.
image

### CODE REVIEW

File: site/app/controllers/admin/AdminGradeableController.php [hunk 1]
```diff
@@ -208,6 +208,16 @@ private function editPage(Gradeable $gradeable, $semester, $course, $nav_tab = 0
             }
         }
         // $this->inherit_teams_list = $this->core->getQueries()->getAllElectronicGradeablesWithBaseTeams();
+        $template_list = $this->core->getQueries()->getAllGradeablesIdsAndTitles();
+
+        $max_points = 0;
+        if ($gradeable->getDependsOn() !== null) {
+            $depend_gradeable = $this->tryGetGradeable($gradeable->getDependsOn());
```
Comment (DefinitiveAbove) [inline site/app/controllers/admin/AdminGradeableController.php:215]: It looks like this function is only being called from a HTML-returning route, so false should be passed in as the second parameter to disable JSON rendering.

File: site/app/templates/admin/admin_gradeable/AdminGradeableAuto.twig [hunk 2]
```diff
@@ -30,6 +30,29 @@
...
+                <select name="depends_on" id="gradeable-lock">
+                    <option value="">Select an option (Off)</option>
+                    {% for template in template_list %}
+                        <option value="{{ template.g_id }}" {% if template.g_id == gradeable.getDependsOn() %} selected {% endif %}>{{ template.g_title }}-{{ template.g_id }}</option>
+                    {% endfor %}
+                </select>
+            </div>
+        </fieldset>
+        <fieldset>
+            <legend>Points Required to Unlock:</legend>
```
Comment (DefinitiveAbove) [inline site/app/templates/admin/admin_gradeable/AdminGradeableAuto.twig:49]: I think that it should be specified here that the points checked are only autograding points and not the overall score.

**Review by DefinitiveAbove — CHANGES_REQUESTED** [review]
Tested and everything looks good functionality wise.

I'm not sure if this was brought up before, but maybe instructors (and full/limited access graders?) should be able to ignore the lock on gradeables for testing submissions.

**Review by DefinitiveAbove — APPROVED** [review]
Looks good to me!

File: site/app/controllers/admin/AdminGradeableController.php [hunk 3]
```diff
@@ -1077,6 +1097,24 @@ private function updateGradeable(Gradeable $gradeable, $details) {
...
 
+            if (in_array($prop, $depend_props)) {
+                if ($post_val === "") {
+                    $post_val = null;
+                }
+            }
+
+            if ($prop === "depends_on") {
+                try {
+                    $gradeables = new GradeableList($this->core);
```
Comment (MasterOdin) [inline site/app/controllers/admin/AdminGradeableController.php:1108]: Why are you constructing a whole GradeableList here and not just doing a simple existence query check?

File: migration/migrator/data/course_tables.sql [hunk 4]
```diff
@@ -201,6 +201,8 @@ CREATE TABLE public.electronic_gradeable (
     eg_peer_blind integer DEFAULT 3,
     eg_grade_inquiry_start_date timestamp(6) with time zone NOT NULL,
     eg_hidden_files character varying(1024),
+    eg_depends_on character varying(255) DEFAULT NULL::character varying,
```
Comment (MasterOdin) [inline migration/migrator/data/course_tables.sql:204]: Looks like `eg_depends_on` should only be set to null or a valid `g_id`. As such, we should add a [foreign key](https://www.postgresqltutorial.com/postgresql-foreign-key/) constraint to ensure data consistency, as well as handling the case where a gradeable is deleted where another gradeable depends on it.

**Review by MasterOdin — CHANGES_REQUESTED** [review]

**Review by eddiekrystowski — CHANGES_REQUESTED** [review]
Seems to work well for the most part, except when I put 0 points required, it always unlocks the next gradeable, even if I haven't made a submission to the prior one yet.

Steps:
1. Create Assignment 1
2. Create Assignment 2
3. Make Assignment 2 depend on Assignment 1 getting 0 points
4. Assignment 2 can now be done even though nothing has been submitted to Assignment 1

**Comment by cjreed121** [comment]
> Seems to work well for the most part, except when I put 0 points required, it always unlocks the next gradeable, even if I haven't made a submission to the prior one yet.
> 
> Steps:
> 
> 1. Create Assignment 1
> 2. Create Assignment 2
> 3. Make Assignment 2 depend on Assignment 1 getting 0 points
> 4. Assignment 2 can now be done even though nothing has been submitted to Assignment 1

@eddiekrystowski Thanks for noticing this. Just a small type error. It seems to work now. Can you give it another try?

**Review by eddiekrystowski — APPROVED** [review]
Works as described. Nice work!

**Comment by bmcutler** [comment]
@cjreed121 Please resolve merge conflicts

**Comment by bmcutler** [comment]
When I test this, 
I'm getting "Internal server error" when choose a gradeable.
Maybe it's unrelated to this PR?  Merge conflicts prevent updating to master branch.

**Comment by bmcutler** [comment]
Maybe the internal server error was related to it being in the a "future" category?  
I wasn't getting the extra info about # of autograding points available -- or maybe I had only picked assignments with no autograding points?

I tried with another gradeable, no internal server error, I did get the message with # of available points, but on leaving the edit gradeable page I got this frog robot message:

Screen Shot 2021-07-07 at 10 53 45 AM

**Review by bmcutler — CHANGES_REQUESTED** [review]
merge conflicts & bugs when tested

**Comment by bmcutler** [comment]
Installation error:  [ DId I only get this because I've tried to test your PR previously?  Or is this a bug? ]

sqlalchemy.exc.ProgrammingError: (psycopg2.errors.DuplicateObject) constraint "fk_depends_on" for relation "electronic_gradeable" already exists

[SQL: ALTER TABLE electronic_gradeable ADD CONSTRAINT fk_depends_on FOREIGN KEY(eg_depends_on) REFERENCES electronic_gradeable(g_id);]

**Comment by cjreed121** [comment]
That would be since it is already installed but I did not but a if not exists check on the constraint. Now it should be there if you pull again.

**Comment by bmcutler** [comment]
it shouldn't let you specify a threshold > max (non extra credit) points:
Screen Shot 2021-07-09 at 10 22 32 AM

**Review by bmcutler — CHANGES_REQUESTED** [review]
I tweaked the text, I'm happy with the UI, but we have a twig bug to fix.

When you make a new gradeable...  and immediately go to the autograding tab of edit gradeable, you get a twig error.  I think the error also comes when the gradeable is rebuilt?  

Screen Shot 2021-07-18 at 11 50 41 PM

Also...  the gradeable is rebuilt even when you don't change the config?  I think we discussed that it should only rebuild when the config is changed.  (If that's not too complicated to pull off.)

**Comment by cjreed121** [comment]
I could not recreate the last issue you mentioned. I know it was an issue earlier in this PR but that should be fixed. Can you try that again with latest code?

**Review by bmcutler — APPROVED** [review]
yay!

**Pull request merged**